crates/nu-protocol/src/io/lines.rs

```diff
@@ -0,0 +1,114 @@
+use crate::{IntoSpanned, ShellError, Span, Value};
+use std::{
+    io::{self, BufRead, BufReader, Read},
+    sync::{atomic::AtomicBool, Arc},
+};
+
+pub struct ByteLines<R: Read>(BufReader<R>);
+
+impl<R: Read> ByteLines<R> {
+    pub fn new(read: R) -> Self {
+        Self(BufReader::new(read))
+    }
+}
+
+impl<R: Read> Iterator for ByteLines<R> {
+    type Item = io::Result<Vec<u8>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut buf = Vec::new();
+        // `read_until` will never stop reading unless `\n` or EOF is encountered,
+        // so we may want to limit the number of bytes by using `take` as the Rust docs suggest.
+        // let capacity = self.0.capacity() as u64;
+        // let mut reader = (&mut self.0).take(capacity);
+        let reader = &mut self.0;
+        match reader.read_until(b'\n', &mut buf) {
+            Ok(0) => None,
+            Ok(_) => {
+                if buf.ends_with(&[b'\n']) {
+                    let _ = buf.pop();
+                    if buf.ends_with(&[b'\r']) {
+                        let _ = buf.pop();
+                    }
+                }
+                Some(Ok(buf))
+            }
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+pub struct Lines<R: Read> {
+    lines: ByteLines<R>,
+    span: Span,
+    ctrlc: Option<Arc<AtomicBool>>,
+}
+
+impl<R: Read> Lines<R> {
+    pub fn new(reader: R, span: Span, ctrlc: Option<Arc<AtomicBool>>) -> Self {
+        Self {
+            lines: ByteLines::new(reader),
+            span,
+            ctrlc,
+        }
+    }
+
+    pub fn span(&self) -> Span {
+        self.span
+    }
+}
+
+impl<R: Read> Iterator for Lines<R> {
+    type Item = Result<Vec<u8>, ShellError>;
```

Comment:
If you're dealing with lines, you're basically always dealing with UTF-8 text anyway (other encodings would potentially have different newline sequences), so in my case I rolled the parsing logic into this iterator and just had it spit out strings. I left `ByteLines` though in case someone really does want that.